### PR TITLE
Add the Compress function (#4)

### PR DESCRIPTION
**Pull request by khnom5000**
Add the Compress function (#4)

### CODE REVIEW

**Comment by khnom5000** [comment]
running over the selector will result in a "panic: runtime error: index out of range", not sure how this wants wants to be handled if at all, left it as default for now

**Comment by khnom5000** [comment]
added a selector len check to stop panic in prev comment, i will also add a benchmark test in

**Review by h-dav — CHANGES_REQUESTED** [review]
Happy with functionality, just a comment about a comment, and a suggestion

File: itertools.go [hunk 1]
```diff
@@ -181,3 +181,16 @@ func ensureSameLength[T any](nestedList [][]T) bool {
 	}
 	return true
 }
+
+func Compress[T any](data []T, selector []bool) (ch Iterator) {
```
[the remark this answers is not in the document]
Comment (khnom5000) [reply itertools.go:185]: will do :)

File: itertools.go [hunk 2]
```diff
@@ -181,3 +181,16 @@ func ensureSameLength[T any](nestedList [][]T) bool {
...
+	ch = make(Iterator)
+	go func() {
+		defer close(ch)
+		for i, d := range data {
+			if len(selector) > i && selector[i] {
```
[the remark this answers is not in the document]
Comment (khnom5000) [reply itertools.go:190]: by doing this it think we then limit ourselves, for example:
if i have data = [1,2,3,4,5,6,7,8,9,0]
and selector = [T,T,F,T]
if i just want values [1,2,4] in the return, then i would need to pad False on the end of the array 6 times or be hit with a check, currently this function would just accept the data and selector above and return [1,2,4].
what do you think?

[the remark this answers is not in the document]
Comment (h-dav) [reply itertools.go:191]: Ah yeah that's a good point. Happy with this now

**Review by h-dav — APPROVED** [review]

**Pull request merged**